libavpipe/src/avpipe_copy_mpegts.c

```diff
@@ -0,0 +1,389 @@
+        elv_err("Failure in preparing audio encoder, rc=%d, url=%s", rc, params->url);
+        return rc;
+    }
+
+    /*
+     * Allocate an array of MAX_STREAMS out_handler_t (a single one for video and audio).
+     * Needs to allocate up to number of streams when transcoding multiple streams at the same time.
+     */
+    out_tracker = (out_tracker_t *) calloc(1, sizeof(out_tracker_t));
+    out_tracker->out_handlers = out_handlers;
+    out_tracker->inctx = inctx;
+    out_tracker->video_stream_index = decoder_context->video_stream_index;
+    out_tracker->audio_stream_index = decoder_context->audio_stream_index[0]; // PENDING(SS) do we need this?
+    out_tracker->seg_index = atoi(params->start_segment_str);
+    out_tracker->encoder_ctx = encoder_context;
+    out_tracker->xc_type = xc_all;
+    encoder_context->format_context->avpipe_opaque = out_tracker;
+
+    return 0;
+}
+
+static int
+copy_mpegts(
+    coderctx_t *decoder_context,
+    coderctx_t *encoder_context,
+    AVPacket *packet,
+    AVFrame *frame,
+    AVFrame *filt_frame,
+    int stream_index,
+    xcparams_t *p,
+    int do_instrument,
+    int debug_frame_level)
+{
+    AVFormatContext *format_context;
+
+    format_context = encoder_context->format_context;
+
+    if (packet->pts == AV_NOPTS_VALUE ||
+        packet->dts == AV_NOPTS_VALUE ||
+        packet->data == NULL) {
+        char *url = "";
+        if (decoder_context->inctx && decoder_context->inctx->url)
+            url = decoder_context->inctx->url;
+        elv_warn("INVALID %s PACKET (COPY) url=%s pts=%"PRId64" dts=%"PRId64" duration=%"PRId64" pos=%"PRId64" size=%d stream_index=%d flags=%x data=%p\n",
+            "AUDIO/VIDEO", url,
+            packet->pts, packet->dts, packet->duration,
+            packet->pos, packet->size, packet->stream_index,
+            packet->flags, packet->data);
+
+        return eav_success; // Respect the logic in regular bypass encoder
+    }
+
+    int rc = av_interleaved_write_frame(format_context, packet);
+    if (rc < 0) {
+        elv_err("Failure in copying packet xc_type=%d rc=%d url=%s", p->xc_type, rc, p->url);
+        return eav_write_frame;
+    }
+
+    // PENDING(SS) call avpipe_stater(outctx)
+
+    return eav_success;
+}
+
+void *
+copy_mpegts_func(
+    void *p)
+{
+    xctx_t *xctx = (xctx_t *) p;
+    cp_ctx_t *cp_ctx = &xctx->cp_ctx;
+
+    coderctx_t *decoder_context = &xctx->decoder_ctx;
+    coderctx_t *encoder_context = &cp_ctx->encoder_ctx;
+    xcparams_t *params = xctx->params;
+    xc_frame_t *xc_frame;
+    int err = 0;
+
+    AVFrame *frame = av_frame_alloc();
+    AVFrame *filt_frame = av_frame_alloc();
+
+    while (!xctx->stop || elv_channel_size(cp_ctx->ch) > 0) {
+
+        xc_frame = elv_channel_receive(cp_ctx->ch);
+        if (!xc_frame) {
+            elv_dbg("copy_mpegts_func, there is no frame, url=%s", params->url);
+            continue;
+        }
```

Comment:
I think it makes sense to make a comment that the `xc_frame` will only contain a packet, and free it here so that it is clear that the rest of it isn't used.

Something like:

```go
// This xc_frame will only have a packet, because it is for directly bypassing to mpegts
xc_frame = elv_channel_receive(cp_ctx->ch);
if (!xc_frame) {
    elv_dbg("copy_mpegts_func, there is no frame, url=%s", params->url);
    continue;
}
AVPacket *packet = xc_frame->packet;
free(xc_frame);

...
```

Reply:
Yes!  Fixed as above.